src/audio/dai-zephyr.c

```diff
@@ -0,0 +1,1408 @@
+		comp_underrun(dev, dd->local_buffer, bytes);
+	} else {
+		comp_err(dev, "dai_report_xrun(): overrun due to no space available");
+		comp_overrun(dev, dd->local_buffer, bytes);
+	}
+}
+
+/* copy and process stream data from source to sink buffers */
+static int dai_copy(struct comp_dev *dev)
+{
+	struct dai_data *dd = comp_get_drvdata(dev);
+	uint32_t dma_fmt = dd->dma_buffer->stream.frame_fmt;
+	const uint32_t sampling = get_sample_bytes(dma_fmt);
+	struct comp_buffer *buf = dd->local_buffer;
+	uint32_t avail_bytes = 0;
+	uint32_t free_bytes = 0;
+	uint32_t copy_bytes = 0;
+	uint32_t src_samples;
+	uint32_t sink_samples;
+	uint32_t samples;
+	int ret = 0;
+
+	comp_dbg(dev, "dai_copy()");
+
+	/* get data sizes from DMA */
+	ret = dma_get_data_size(dd->chan, &avail_bytes, &free_bytes);
+	if (ret < 0) {
+		dai_report_xrun(dev, 0);
+		return ret;
+	}
+
+	buf = buffer_acquire(buf);
+
+	/* calculate minimum size to copy */
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK) {
+		src_samples = audio_stream_get_avail_samples(&buf->stream);
+		sink_samples = free_bytes / sampling;
+		samples = MIN(src_samples, sink_samples);
+	} else {
+		src_samples = avail_bytes / sampling;
+		sink_samples = audio_stream_get_free_samples(&buf->stream);
+		samples = MIN(src_samples, sink_samples);
+	}
+
+	/* limit bytes per copy to one period for the whole pipeline
+	 * in order to avoid high load spike
+	 */
+	samples = MIN(samples, dd->period_bytes / sampling);
+
+	copy_bytes = samples * sampling;
+
+	buffer_release(buf);
+
+	comp_dbg(dev, "dai_copy(), dir: %d copy_bytes= 0x%x, frames= %d",
+		 dev->direction, copy_bytes,
+		 samples / buf->stream.channels);
+
+	/* Check possibility of glitch occurrence */
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK &&
+	    copy_bytes + avail_bytes < dd->period_bytes)
+		comp_warn(dev, "dai_copy(): Copy_bytes %d + avail bytes %d < period bytes %d, possible glitch",
+			  copy_bytes, avail_bytes, dd->period_bytes);
+	else if (dev->direction == SOF_IPC_STREAM_CAPTURE &&
+		 copy_bytes + free_bytes < dd->period_bytes)
+		comp_warn(dev, "dai_copy(): Copy_bytes %d + free bytes %d < period bytes %d, possible glitch",
+			  copy_bytes, free_bytes, dd->period_bytes);
+
+	/* return if nothing to copy */
+	if (!copy_bytes) {
+		comp_warn(dev, "dai_copy(): nothing to copy");
+		return 0;
+	}
+
+	ret = dma_copy(dd->chan, copy_bytes, 0);
+	if (ret < 0) {
+		dai_report_xrun(dev, copy_bytes);
+		return ret;
+	}
+
+	dai_dma_position_update(dev);
+
+	return ret;
+}
+
+/**
+ * \brief Get DAI parameters and configure timestamping
+ * \param[in, out] dev DAI device.
+ * \return Error code.
+ *
+ * This function retrieves various DAI parameters such as type, direction, index, and DMA
+ * controller information those are needed when configuring HW timestamping. Note that
+ * DAI must be prepared before this function is used (for DMA information). If not, an error
+ * is returned.
+ */
+static int dai_ts_config_op(struct comp_dev *dev)
+{
+#if 0
```

Comment:
timestamp should be upstream in Zephyr as these APIs are for IP programming.

Reply:
ok will call zephyr dai methods